llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1207,6 +1207,45 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
     }
   }
 
+  // (X * Y) / (X * Z) --> Y / X (and commuted variants)
+  if (match(Op0, m_Mul(m_Value(X), m_Value(Y)))) {
+    auto OB0HasNSW = cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap();
+    auto OB0HasNUW = cast<OverflowingBinaryOperator>(Op0)->hasNoUnsignedWrap();
+
+    auto CreateDivOrNull = [&](Value *A, Value *B) -> Instruction * {
+      auto OB1HasNSW = cast<OverflowingBinaryOperator>(Op1)->hasNoSignedWrap();
+      auto OB1HasNUW =
+          cast<OverflowingBinaryOperator>(Op1)->hasNoUnsignedWrap();
+      const APInt *C;
+      if (IsSigned && OB0HasNSW && OB1HasNSW && (match(A, m_APInt(C))) &&
+          (*C != -1 ||
+           *C != APInt::getSignedMinValue(Ty->getScalarSizeInBits())))
+        return BinaryOperator::CreateSDiv(A, B);
+      if (!IsSigned && OB0HasNUW && OB1HasNUW) {
+        return BinaryOperator::CreateUDiv(A, B);
+      };
+      if (isa<Constant>(A) && isa<Constant>(B)) {
+        if (IsSigned && OB0HasNSW) {
+          return BinaryOperator::CreateSDiv(A, B);
+        }
+        if (!IsSigned && OB0HasNUW) {
+          return BinaryOperator::CreateUDiv(A, B);
+        };
+      };
```

Comment:
I think it may cause miscompilation. I believe it is safe iff `A / B` is exact.
Please provide alive2 proof for it.

Reply:
Alive2: https://alive2.llvm.org/ce/z/2gkqu7

Reply:
Maybe this is better? https://alive2.llvm.org/ce/z/HL3K6G